move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h

robieta opened the pull request:
There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59393
* #59392
* **#59391**
* #59389

Commit:
move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h

[ghstack-poisoned]

swolchok commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -24,67 +25,45 @@
+// convention and the "normal" in memory layout.
+class C10_API LocalDispatchKeySetWrapper {
+ public:
+  LocalDispatchKeySetWrapper(PODLocalState* tls) : tls_(tls) {}
```
explicit?

robieta replied:
Done.

swolchok commented on c10/core/impl/ThreadLocalState.cpp:
```diff
@@ -0,0 +1,16 @@
+namespace c10 {
+namespace impl {
+
+// static_assert(false, "test");
```
delete

robieta replied:
Done.

ezyang commented:
You mentioned briefly in work chat about why this is profitable but it's not written on the PR?

ezyang commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -320,7 +320,7 @@ Operation BroadOp(const Node* node) {
         // We use a very simple heuristic to convert an arg in nchw
         // to the blocked format of the other argument.
-        c10::impl::ExcludeDispatchKeyGuard edkg(c10::autograd_dispatch_keyset);
+        c10::impl::ExcludeDispatchKeyGuard_NoOverlap<c10::autograd_dispatch_keyset.raw_repr()> edkg;
```
Why do we not just have a `ExcludeAutogradDispatchKeyGuard`

robieta replied:
This was because autocast does overlap. I realized I can coerce these into one manager, though I have to get the autocast keys to be constexpr. (Which they really should be in the first place.)

ezyang commented:
> This PR pulls the state of the local dispatch key sets into a separate TLS. This is broken into a separate PR so that we can check that the instruction count is unchanged. (Which it is.) Future PRs in the stack will add more fields to this TLS.

If you could say more about the overall plan that would also be helpful

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -24,67 +25,45 @@
 
-  void set_included(DispatchKeySet x) {
-    included_ = (x ^ c10::default_included_set).raw_repr();
+  void unsafe_set_included(DispatchKeySet x) {
```
It behooves you to explain why these are unsafe

robieta replied:
I don't actually remember. I think it might have been from an earlier iteration where I had dropped the mask. (I'll drop the name now.)

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -24,67 +25,45 @@
-  uint64_t excluded_;
+// Helper class to handle conversion between the TLS zero initialized
+// convention and the "normal" in memory layout.
+class C10_API LocalDispatchKeySetWrapper {
```
@swolchok I guess, making a class here isn't a pessimization? I was worried about the class members being forced to put into memory but I tried a few simple and gcc was able to eliminate the class entirely.

robieta replied:
Yeah, I checked that it's identical.

swolchok replied:
as long as all the member functions are inlineable, you should be OK.

ezyang commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -36,12 +36,7 @@ static inline DispatchKeySet computeDispatchKeySet(
-  // be nice to only do one.  Can always_included be folded into the TLS?  Well,
-  // it's a bit troublesome, because fastpath TLS access requires the type of
-  // the TLS in question to be zero-initialized, so you don't actually win
-  // anyting in that case.
```
Is this comment obsolete? It doesn't look obsolete to me.

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -24,67 +25,45 @@
-  uint64_t included_;
-  uint64_t excluded_;
+// Helper class to handle conversion between the TLS zero initialized
+// convention and the "normal" in memory layout.
```
It's not just this right, idiomatic usage seems to involve the default constructor, which does a TLS fetch. (To be honest, I'd prefer it if it were more obvious when a TLS fetch was happening, rather than doing so implicitly)

robieta replied:
SGTM

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

This PR pulls the state of the local dispatch key sets into a separate TLS. This is broken into a separate PR so that we can check that the instruction count is unchanged. (Which it is.) Future PRs in the stack will add more fields to this TLS.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

This PR pulls the state of the local dispatch key sets into a separate TLS. This is broken into a separate PR so that we can check that the instruction count is unchanged. (Which it is.) Future PRs in the stack will add more fields to this TLS.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

This PR pulls the state of the local dispatch key sets into a separate TLS. This is broken into a separate PR so that we can check that the instruction count is unchanged. (Which it is.) Future PRs in the stack will add more fields to this TLS.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

This PR pulls the state of the local dispatch key sets into a separate TLS. This is broken into a separate PR so that we can check that the instruction count is unchanged. (Which it is.) Future PRs in the stack will add more fields to this TLS.



[ghstack-poisoned]

ailzhang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -113,10 +102,51 @@ class C10_API ExcludeDispatchKeyGuard {
+template <uint64_t exclude>
+class C10_API ExcludeDispatchKeyGuard_NoOverlap {
+ public:
+  // If our exclude set does not overlap with c10::default_excluded_set, we can
```
Hmmm please correct me if I'm wrong: if I want to add a key into c10::default_excluded_set in the future, do I have to check all call sites of `ExcludeDispatchKeyGuard_NoOverlap` and make sure they still don't overlap? ;) If so this seems a bit hard to maintain compared to the previous version.

robieta replied:
I'm merging both paths into a single guard. (Where you declare if there is overlap) I could make it automatic, but there aren't that many guards and I don't want a change in default_excluded_set to silently regress perf by kicking us off the hot path.

ezyang replied:
I guess, what is unclear to me even after reading the explanation here, is if I change `default_excluded_set` and this static assert fires, what the heck am I supposed to do in that case? Am I supposed to edit the `has_overlap` arguments (won't that regress perf?). But then I am confused what else I am supposed to do. Is this a case of there being duplicated information that has to be kept up to date in multiple places? If that's the case, can we clearly indicate where else things have to be updated (or make it so that an update is not necessary in the first place?)

robieta replied:
Yes, the idea is that you have to update `has_overlap`; effectively forcing a discussion in code review about whether it's ok to knock that guard off the hot path. I can deduce `has_overlap`, it just means we can have silent perf regressions.

robieta commented on c10/core/impl/ThreadLocalState.h:
```diff
@@ -0,0 +1,43 @@
+#pragma once
```
I actually picked the name before I knew about the aten one. Bikeshedding welcome.

ezyang replied:
Well, inside of this file, the struct is named PODLocalState, so that seems like a much better name for this file.

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

This PR pulls the state of the local dispatch key sets into a separate TLS. This is broken into a separate PR so that we can check that the instruction count is unchanged. (Which it is.) Future PRs in the stack will add more fields to this TLS.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

This PR pulls the state of the local dispatch key sets into a separate TLS. This is broken into a separate PR so that we can check that the instruction count is unchanged. (Which it is.) Future PRs in the stack will add more fields to this TLS.



[ghstack-poisoned]

ilia-cher commented:
just to confirm - is the PODLocalState used for dispatch only? if yes, could we make it more obvious to avoid confusion with generic (and non-POD) ThreadLocalState (including the file names).

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -24,57 +27,44 @@
-  // Don't let people fiddle with the thread_local directly just
-  // because they include this header.
-  return raw_local_dispatch_key_set;
+inline C10_API LocalDispatchKeySet snapshot_tls_keyset() {
```
I wonder what the "snapshot" here is supposed to mean. (Since you've gone through the trouble of renaming it I assume there's some motive here, but I can't figure it out from the code.) In particular, "snapshot" sounds like a verb but this function has no side effects.

ezyang replied:
I guess you intend this to mean, "take a snapshot of the current tls keyset"? But that's a weird way to say it; if I define a getter for an integer field foobar on a class that I'm about to change, I don't call it `snapshot_foobar`, I just say `foobar` or `get_foobar`...

robieta replied:
The key point that I'm trying to convey is that this copies the data in the TLS rather than returning a view. So mutating the LocalDispatchKeySet that is returned does not propagate to the source state. But if that is confusing things more than helping I'm happy to go back to `tls_local_dispatch_key_set()`.

bhosmer replied:
I think `tls_keyset_snapshot` would do it. Also though, re "mutating does not propagate", why not just make LocalDispatchKeySet members immutable? That would take some of the weight off a name change - `tls_keyset_snapshot` is arguably more descriptive, but the diff shed some jitter if we kept the current one. 

Aside: I'm hopping around these PRs thinking about your keep-or-ditch question, so I realize this comment won't necessarily be useful - but just in case 😁

robieta replied:
I experimented with making the members `const`, but it broke certain builds unfortunately. (It wasn't clear what was going on so I didn't pursue it further.)

ezyang replied:
In regular C++, if I have a function that returns a struct by value, and the struct is not obviously named like a reference (e.g., `ArrayRef` or `StringView`), then most reasonable readers will assume that value semantics apply here (in the same way that `std::vector<int> f()` gives you a fresh copy of the vector, it never aliases with anything else).

PyTorch blurs the lines here by defining `Tensor` to be a reference type. But in that case, `clone_tls_keyset()` is a good, idiomatic naming for this codebase, since clone is the name of the corresponding operation on tensors that doesn't give you an alias but in fact a new copy. I still maintain snapshot is totally unclear.

bhosmer replied:
Personally I don't think its meaning is unclear, but agreed that it's not a modifier in common use in the codebase. TBH I think the old name is fine, and would avoid some churn.

bhosmer replied:
> I experimented with making the members `const`, but it broke certain builds unfortunately. (It wasn't clear what was going on so I didn't pursue it further.)

It's probably worth a quick check of the errors to see if they're "real" in the sense that people are actually depending on the stateful updating of `LocalDispatchKeySet`'s properties, as opposed to just doing after-construction initialization, e.g. see my comment on the code in `InferenceMode`'s ctor.

ilia-cher commented on aten/src/ATen/ThreadLocalState.cpp:
```diff
@@ -9,7 +9,7 @@
 
 ThreadLocalState::ThreadLocalState(bool keep_grad_mode)
-    : dispatch_key_(c10::impl::tls_local_dispatch_key_set()),
+    : dispatch_key_(c10::impl::snapshot_tls_keyset()),
```
given that you add grad mode later into the new TLS state, should we copy the whole POD state here?

robieta replied:
That kind of already happens with the `grad_mode_enabled_ = GradMode::is_enabled();` call. The only difference is that the source of dispatch keys and grad mode is the same TLS instead of two different ones. The more I look at this the less thrilled I am with the interaction of all the TLS's. (This PR isn't making it worse, it's just highlighting that we have kind of a lot of duplicated state.)

robieta replied:
Although that wouldn't respect `keep_grad_mode`.

ilia-cher replied:
oh, right, re. `keep_grad_mode` - the only use case for this is to avoid enabling grad mode within autograd thread itself, when transitioning from the forward pass into the backward pass

ezyang commented:
I'm going to repeat myself from six days ago: **moar PR description please.** If you want some examples of what level of description I'm looking for, checkout some of my recent PRs: https://github.com/pytorch/pytorch/pull/59760 https://github.com/pytorch/pytorch/pull/57985 https://github.com/pytorch/pytorch/pull/55799

robieta commented:
> I'm going to repeat myself from six days ago: **moar PR description please.** If you want some examples of what level of description I'm looking for, checkout some of my recent PRs: #59760 #57985 #55799

I saw this. I am midway through refactoring the stack and was going to update the descriptions before requesting that folks take another look, but I'll do it now.

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

Commit:
Update on "move local dispatch key storage to a TLS seperate from the LocalDispatchKeySet.h"

There are currently four TLS lookups on the autograd dispatch key path:
1) A lookup to fetch the local dispatch key set for initial call
2) A lookup to check the RecordFunction state to determine if we need to record the call
3) A lookup inside the autograd kernel to check if grad mode is active
4) A lookup of the same keyset from (1) for AutoDispatchBelowADInplaceOrView

If we coalesce thread local state into a single struct, we can reuse the pointer and skip TLS address lookups. This is complicated, however, by the requirement that this thread local data be plain old data (POD). This PR factors the dispatch key state into a standalone TLS struct and adapts the local dispatch key machinery to use external storage (the central TLS); this means that if one already has the TLS pointer on hand operations on the local dispatch key set can skip the lookup work. Subsequent PRs in the stack add additional fields and replicate the pattern. The reason I started with a degenerate (single field)  case was to validate that separating the storage and accessor did not impact perf.

There is an additional finesse in this PR. Due to the POD nature of the TLS, we have to play some XOR games with the storage. However in common special cases (which are known at compile time) there is no bit overlap and we can skip the XOR conversion and still maintain correctness.



[ghstack-poisoned]

robieta commented:
OMG TorchVision uses `c10::impl::ExcludeDispatchKeyGuard`...

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -95,28 +85,85 @@ class C10_API IncludeDispatchKeyGuard {
+// Create a guard which can be exported by subclassing a specialization
+// of ExcludeDispatchKeyGuard. Note that because it will be created in ATen,
+// we need TORCH_API rather than C10_API.
+#define SPECIALIZE_EXCLUDE_GUARD(guard_name, k, has_overlap)              \
```
Macros in headers need to be appropriately namespaced; e.g., `TORCH_SPECIALIZE_EXCLUDE_GUARD`

ezyang replied:
(or `C10_SPECIALIZE_EXCLUDE_GUARD`, which may be more appropriate given this is in c10/ folder)

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -107,6 +107,9 @@ class DispatchKeySet final {
   }
+  constexpr uint64_t raw_repr() const {
+    return repr_;
+  }
```
There's reason to add a const overload here; the const variant is also valid in contexts where you have a mutable ref. Just delete the old accessor on line 107

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -24,57 +27,44 @@
-  uint64_t excluded_;
+// Helper class to handle conversion between the TLS zero initialized
+// convention and the "normal" in memory layout.
+class C10_API LocalDispatchKeySetWrapper {
```
The naming scheme here is a bit confusing. Ordinarily, I would expect a LocalDispatchKeySetWrapper to be something that wraps a LocalDispatchKeySet. But actually, this is a wrapper on PODLocalState which makes it look like LocalDispatchKeySet (but there is still a LocalDispatchKeySet which will copy the POD local state out).

To kick off the bikeshed, I suggest renaming the wrapper classes to something like "PODAsLocalDispatchKeySet" (or "PODLocalStateAsLocalDispatchKeySet" if you want to be super wordy about it)

ezyang replied:
Another consideration here is that this is a non-owning pointer to the underlying POD state, but you still use this as a normal field inside other structs. It's extra important to emphasize that this struct aliases into the true POD state (that other TLS fetches will see), rather than being just another copy of the state (which is the semantics of LocalDispatchKeyState). So if you wanted a different naming scheme, you might also consider `LocalDispatchKeySetTLSHandle` -- emphasizing the fact that it is a handle to TLS state)

bhosmer replied:
I think the whole setup can be simplified/clarified by making `LocalDispatchKeySetWrapper` *always* hold a pointer to `_get_thread_local_state()` - i.e., give it nothing but a default constructor that does this initialization - and renaming the class accordingly, e.g. `LocalDispatchKeySetTLSHandle` per Ed's suggestion, or something similar. (Even `ThreadLocalDispatchKeySet`.) 
 
Making it so an instance of this class is by definition a facade on the TLS representation will really clarify usage, I think. (I couldn't find any ctor call sites that initialized with anything else, but if "untethered" use cases exist I'd argue from separation of concerns that we should have a different class for them.)

robieta replied:
I think this is reasonable. The reason for the ctr that takes a pointer is that at one point I was plumbing the tls pointer along with the dispatch keyset to save a TLS lookup. But it turned out not to be worth it to enlarge the signature. So yeah, at this point there's no reason for the decoupling. (Which also makes the `Handle` naming make a lot of sense.)

Thanks for the discussion; this is really great feedback and I think will simplify things a lot.

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -24,57 +27,44 @@
 
 struct C10_API LocalDispatchKeySet {
-  /* implicit */ LocalDispatchKeySet(PODLocalDispatchKeySet x)
+  LocalDispatchKeySet(LocalDispatchKeySetWrapper x)
```
this is still implicit, fwiw

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -95,28 +85,85 @@ class C10_API IncludeDispatchKeyGuard {
+  // declared so that changes to the default excluded set do not silently knock
+  // us off the hot path.
+  static_assert(
+      has_overlap == (bool)(exclude & c10::default_excluded_set.raw_repr()),
```
nit: `static_cast<bool>`

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -95,28 +85,85 @@ class C10_API IncludeDispatchKeyGuard {
+  // Key exclusion tends to be on the hot path, so it's worth it to bypass the
+  // unnecessary XORs if possible. We force `has_overlap` to be explicitly
+  // declared so that changes to the default excluded set do not silently knock
+  // us off the hot path.
```
I'm not convinced by the explicit `has_overlap` argument here. Suppose that someone makes a change to default excluded set, and it is valid to remove things off the hot path. Is the point here to just to force someone to notice, in code review, that we are no longer on the hot path? It seems like you could do that a lot more straightforwardly by just having a sequence of static asserts that such and such dispatch key is in (not) in the default excluded set. This would still have the desired path of flagging such a change in review... while not simultaneously forcing the person to also change all of the call sites to update the (redundant) has overlap information.

bhosmer replied:
+1, I think this encumbers the API a bit too much, in addition to making changes potentially higher friction than they need to be. Taylor do you see a problem with going with Ed's suggestion of just static asserting the keys we care about directly instead?

robieta replied:
Yeah, I think that is very reasonable. I think this also lets us get rid of the specialize macro altogether.

bhosmer commented on c10/core/InferenceMode.h:
```diff
@@ -60,9 +60,9 @@ struct TORCH_API InferenceMode {
-    cur_keyset.set_excluded(excluded);
+    auto cur_keyset = prev_keyset;
+    cur_keyset.included_ = included;
+    cur_keyset.excluded_ = excluded;
```
Note that what's happening here is `cur_keyset` is being initialized from `prev_keyset`, then both (i.e., all) of its properties are being updated. Adding another ctor to `LocalDispatchKeySet` which took DispatchKeySets for `included_` and `excluded_` directly would remove the need for them to be non-const here. (Don't know if there are other places but I imagine they could be dealt with similarly.)

At a high level, the pitch is that by making `LocalDispatchKeySet` axiomatically immutable, and making (the class currently known as) `LocalDispatchKeySetWrapper` axiomatically a facade on TLS, everything's a bit cleaner and easier to reason about - e.g. you don't need to work as hard to figure out whether a property update is gonna trickle back into TLS or not.

bhosmer reviewed (approved):
Thanks for your patience Taylor - got a couple suggestions inline, LMK if they make sense, happy to follow up offline :)

bhosmer reviewed (commented):
Whoops, sorry, didn't mean to approve just yet.

bhosmer reviewed (changes requested):
Just unapproving my inadvertent approve - there's enough still to talk through here to warrant keeping the green check available as a phase change marker.

ezyang commented on c10/core/impl/ThreadLocalState.h:
```diff
@@ -0,0 +1,42 @@
+//
+// Moreover, PODLocalState is a simple data container. Logic for interpreting
+// or updating fields should be owned by higher level containers which operate
+// on a PODLocalState* pointer.
```
Since it's hard coded here what PODLocalState actually contains, there is only a fixed number of higher level containers. Help the reader out by giving direct code pointers to the full set of higher level containers.

bhosmer reviewed (commented):
Chiming in on a couple more things, nothing huge but see inline
